File: python/src/workshop_api/external_invoice_provider/invoice_provider_controller.py
import json
import os
import uuid
from datetime import UTC, datetime

import httpx
from fastapi import APIRouter, Depends, HTTPException, Request, Response, status

from workshop_api.external_invoice_provider.invoice_provider_schemas import (
    InvoiceProviderPaymentReceivedCallbackRequest,
    InvoiceProviderResponse,
    InvoiceProviderUpsertRequest,
)
from workshop_api.external_invoice_provider.invoice_provider_status import (
    InvoiceProviderStatus,
)
from workshop_api.external_invoice_provider.invoice_provider_store import (
    InvoiceProviderStore,
)

router = APIRouter(
    prefix="/api/shared/external-invoice-provider/invoices",
    tags=["external-invoice-provider"],
)

store = InvoiceProviderStore()


def get_fitness_api_base_url() -> str:
    return os.getenv("WORKSHOP_FITNESS_API_BASE_URL", "http://127.0.0.1:9090")


@router.get("", response_model=list[InvoiceProviderResponse], response_model_by_alias=True)
def list_invoices() -> list[InvoiceProviderResponse]:
    return store.find_all()


@router.get(
    "/{invoice_id}",
    response_model=InvoiceProviderResponse,
    response_model_by_alias=True,
)
def get_invoice(invoice_id: str) -> InvoiceProviderResponse:
    invoice = store.find_by_id(invoice_id)
    if invoice is None:
        raise _not_found(invoice_id)
    return invoice


@router.post(
    "",
    response_model=InvoiceProviderResponse,
    response_model_by_alias=True,
    status_code=status.HTTP_201_CREATED,
)
def create_invoice(
    request: InvoiceProviderUpsertRequest,
    response: Response,
) -> InvoiceProviderResponse:
    invoice_id = str(uuid.uuid4())
    created = store.save(invoice_id, request)
    response.headers["Location"] = f"/api/shared/external-invoice-provider/invoices/{invoice_id}"
    return created


@router.post(
    "/{invoice_id}/mark-paid",
    response_model=InvoiceProviderResponse,
    response_model_by_alias=True,
)
async def mark_invoice_paid(
    invoice_id: str,
    request: Request,
    fitness_api_base_url: str = Depends(get_fitness_api_base_url),
) -> InvoiceProviderResponse:
    invoice = store.find_by_id(invoice_id)
    if invoice is None:
        raise _not_found(invoice_id)

    if invoice.status == InvoiceProviderStatus.PAID:
        return invoice

    paid_invoice = store.save_response(
        InvoiceProviderResponse(
            invoiceId=invoice.invoice_id,
            customerReference=invoice.customer_reference,
            contractReference=invoice.contract_reference,
            amountInCents=invoice.amount_in_cents,
            currency=invoice.currency,
            dueDate=invoice.due_date,
            status=InvoiceProviderStatus.PAID,
            description=invoice.description,
            externalCorrelationId=invoice.external_correlation_id,
            metadata=invoice.metadata,
        )
    )

    callback_request = InvoiceProviderPaymentReceivedCallbackRequest(
        externalInvoiceId=paid_invoice.invoice_id,
        externalInvoiceReference=paid_invoice.external_correlation_id,
        membershipId=paid_invoice.contract_reference,
        paidAt=datetime.now(UTC),
    )

    client_kwargs: dict[str, object] = {"base_url": fitness_api_base_url}
    if fitness_api_base_url.startswith("http://testserver"):
        client_kwargs["transport"] = httpx.ASGITransport(app=request.app)

    try:
        async with httpx.AsyncClient(**client_kwargs) as client:
            callback_response = await client.post(
                "/api/memberships/payment-received",
                json=callback_request.model_dump(by_alias=True, mode="json")
                if hasattr(callback_request, "model_dump")
                else json.loads(callback_request.json(by_alias=True)),
            )
            callback_response.raise_for_status()
    except httpx.HTTPError:
        pass

    return paid_invoice


@router.put(
    "/{invoice_id}",
    response_model=InvoiceProviderResponse,
    response_model_by_alias=True,
)
def update_invoice(
    invoice_id: str,
    request: InvoiceProviderUpsertRequest,
) -> InvoiceProviderResponse:
    if store.find_by_id(invoice_id) is None:
        raise _not_found(invoice_id)
    return store.save(invoice_id, request)


@router.delete("/{invoice_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_invoice(invoice_id: str) -> Response:
    if store.find_by_id(invoice_id) is None:
        raise _not_found(invoice_id)
    store.delete(invoice_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


def _not_found(invoice_id: str) -> HTTPException:
    return HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail=f"External invoice {invoice_id} was not found",
    )

File: java/src/main/java/com/workshop/architecture/external_invoice_provider/InvoiceProviderController.java
package com.workshop.architecture.external_invoice_provider;

import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.client.RestClient;
import org.springframework.web.client.RestClientException;
import org.springframework.web.server.ResponseStatusException;

import java.net.URI;
import java.time.Instant;
import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/api/shared/external-invoice-provider/invoices")
public class InvoiceProviderController {

    private final InvoiceProviderStore store;
    private final RestClient restClient;

    public InvoiceProviderController(
            InvoiceProviderStore store,
            RestClient.Builder restClientBuilder,
            @Value("${workshop.fitness-api.base-url}") String fitnessApiBaseUrl
    ) {
        this.store = store;
        this.restClient = restClientBuilder.baseUrl(fitnessApiBaseUrl).build();
    }

    @GetMapping
    List<InvoiceProviderResponse> listInvoices() {
        return store.findAll();
    }

    @GetMapping("/{invoiceId}")
    InvoiceProviderResponse getInvoice(@PathVariable String invoiceId) {
        return store.findById(invoiceId)
                .orElseThrow(() -> notFound(invoiceId));
    }

    @PostMapping
    ResponseEntity<InvoiceProviderResponse> createInvoice(
            @Valid @RequestBody InvoiceProviderUpsertRequest request
    ) {
        String invoiceId = UUID.randomUUID().toString();
        InvoiceProviderResponse response = store.save(invoiceId, request);
        return ResponseEntity
                .created(URI.create("/api/shared/external-invoice-provider/invoices/" + invoiceId))
                .body(response);
    }

    @PostMapping("/{invoiceId}/mark-paid")
    InvoiceProviderResponse markInvoicePaid(@PathVariable String invoiceId) {
        InvoiceProviderResponse invoice = store.findById(invoiceId)
                .orElseThrow(() -> notFound(invoiceId));

        if (invoice.status() == InvoiceProviderStatus.PAID) {
            return invoice;
        }

        InvoiceProviderResponse paidInvoice = store.save(new InvoiceProviderResponse(
                invoice.invoiceId(),
                invoice.customerReference(),
                invoice.contractReference(),
                invoice.amountInCents(),
                invoice.currency(),
                invoice.dueDate(),
                InvoiceProviderStatus.PAID,
                invoice.description(),
                invoice.externalCorrelationId(),
                invoice.metadata()
        ));

        try {
            restClient.post()
                    .uri("/api/memberships/payment-received")
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(new InvoiceProviderPaymentReceivedCallbackRequest(
                            paidInvoice.invoiceId(),
                            paidInvoice.externalCorrelationId(),
                            paidInvoice.contractReference(),
                            Instant.now()
                    ))
                    .retrieve()
                    .toBodilessEntity();
        } catch (RestClientException ignored) {
            // Step 2 keeps payment simulation usable even if the callback receiver is not live yet.
        }

        return paidInvoice;
    }

    @PutMapping("/{invoiceId}")
    InvoiceProviderResponse updateInvoice(
            @PathVariable String invoiceId,
            @Valid @RequestBody InvoiceProviderUpsertRequest request
    ) {
        if (store.findById(invoiceId).isEmpty()) {
            throw notFound(invoiceId);
        }
        return store.save(invoiceId, request);
    }

    @DeleteMapping("/{invoiceId}")
    ResponseEntity<Void> deleteInvoice(@PathVariable String invoiceId) {
        if (store.findById(invoiceId).isEmpty()) {
            throw notFound(invoiceId);
        }
        store.delete(invoiceId);
        return ResponseEntity.noContent().build();
    }

    private ResponseStatusException notFound(String invoiceId) {
        return new ResponseStatusException(
                HttpStatus.NOT_FOUND,
                "External invoice %s was not found".formatted(invoiceId)
        );
    }
}
